src/link/script.y

```diff
@@ -0,0 +1,479 @@
+	} else if (addr > endaddr(activeType)) { // Allow "one past the end" sections.
+		scriptError(context, "Cannot go to $%04" PRIx32 ": %s ends at $%04" PRIx16 "",
+		            addr, typeInfo.name.c_str(), endaddr(activeType));
+		pc = endaddr(activeType);
+	} else {
+		pc = addr;
+	}
+}
+
+static void alignTo(uint32_t alignment, uint32_t alignOfs) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+	auto &pc = curAddr[activeType][activeBankIdx];
+
+	// TODO: maybe warn if truncating?
+	alignOfs %= 1 << alignment;
+
+	assert(pc >= typeInfo.startAddr);
+	uint16_t length = alignment < 16 ? (uint16_t)(alignOfs - pc) % (1u << alignment)
+	                                 : alignOfs - pc; // Let it wrap around, this'll trip the check.
+	if (uint16_t offset = pc - typeInfo.startAddr; length > typeInfo.size - offset) {
+		scriptError(context, "Cannot align: the next suitable address after $%04" PRIx16 " is $%04" PRIx16 ", past $%04" PRIx16,
+		            pc, (uint16_t)(pc + length), (uint16_t)(endaddr(activeType) + 1));
+	} else {
+		pc += length;
+	}
+}
+
+static void pad(uint32_t length) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+	auto &pc = curAddr[activeType][activeBankIdx];
+
+	assert(pc >= typeInfo.startAddr);
+	if (uint16_t offset = pc - typeInfo.startAddr; length + offset > typeInfo.size) {
+		scriptError(context, "Cannot pad by %u bytes: only %u bytes to $%04" PRIx16,
+		            length, typeInfo.size - offset, (uint16_t)(endaddr(activeType) + 1));
+	} else {
+		pc += length;
+	}
+}
+
+static void placeSection(std::string const &name) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+
+	// A type *must* be active.
+	if (activeType == SECTTYPE_INVALID) {
+		scriptError(context, "No memory region has been specified to place section \"%s\" in",
+		            name.c_str());
+		return;
+	}
+
+	auto *section = sect_GetSection(name.c_str());
+	if (!section) {
+		scriptError(context, "Unknown section \"%s\"", name.c_str());
+		return;
+	}
+
+	assert(section->offset == 0);
+	// Check that the linker script doesn't contradict what the code says.
+	if (section->type == SECTTYPE_INVALID) {
+		// SDCC areas don't have a type assigned yet, so the linker script is used to give them one.
+		for (Section *fragment = section; fragment; fragment = fragment->nextu) {
+			fragment->type = activeType;
+		}
+	} else if (section->type != activeType) {
+		scriptError(context, "\"%s\" is specified to be a %s section, but it is already a %s section",
+		            name.c_str(), typeInfo.name.c_str(), sectionTypeInfo[section->type].name.c_str());
+	}
+
+	uint32_t bank = activeBankIdx + typeInfo.firstBank;
+	if (section->isBankFixed && bank != section->bank) {
+		scriptError(context, "Linker script wants section \"%s\" to go to %s bank %" PRIu32 ", but it is already in bank %" PRIu32,
```

Comment:
For symmetry with the above `section->type != activeType`, and do avoid saying the linker script "wants" something:
```suggestion
		scriptError(context, "\"%s\" is specified to go in %s bank %" PRIu32 ", but it is already in bank %" PRIu32,
```

Reply:
The problem with a more generic "is specified to" wording, is that it's unclear which of the two sources of truth the message is referring to... I wanted to explicitly point out which constraint comes from the linker script, and using passive voice was *horrible*.

Reply:
You are agreed with about the passive voice. ;) (Although it may be worth updating other error messages, in RGBASM as well, which use it.) I do still want to avoid claiming that RGBLINK "wants" things; maybe "`The linker script places section "%s" in bank %d, but it was already defined in bank %d`" (its definition site in asm being where the conflicting bank comes from). Likewise for the rest.